File: sdk/src/events.ts
/**
 * Event parser for UTXOpia sol_log_data events
 *
 * Events are emitted by the on-chain program as base64-encoded log data.
 * Transaction logs contain lines like: "Program data: <base64>"
 * Each base64 segment decodes to one slice from sol_log_data.
 *
 * ## Events
 *
 * - 0x02 NullifierSpent: disc(1) + hash(32) + op_type(1) = 34 bytes
 * - 0x03 StealthAnnouncement: disc(1) + type(1) + ephemeral(32) + amount(8) + commitment(32) + leaf_index(4) = 78 bytes
 * - 0x0B NullifiersBatch: flat payload in single segment
 * - 0x0C AnnouncementsBatch: flat payload in single segment
 */

/** Event discriminators matching contracts/programs/utxopia/src/utils/events.rs */
export const EVENT_NULLIFIER_SPENT = 0x02;
export const EVENT_STEALTH_ANNOUNCEMENT = 0x03;
export const EVENT_NULLIFIERS_BATCH = 0x0b;
export const EVENT_ANNOUNCEMENTS_BATCH = 0x0c;
/** Phase 2: sender memo (XChaCha20-Poly1305 AEAD payload). */
export const EVENT_SENDER_MEMO = 0x12;
export const EVENT_BTC_ORIGIN_ATTESTATION = 0x15;

/** Parsed nullifier spent event */
export interface NullifierSpentEvent {
  type: "nullifier_spent";
  nullifierHash: Uint8Array; // 32 bytes
  operationType: number;
}

/** Parsed stealth announcement event (includes token_id) */
export interface StealthAnnouncementEvent {
  type: "stealth_announcement";
  announcementType: number; // 0=deposit, 1=transfer
  ephemeralPub: Uint8Array; // 32 bytes
  encryptedAmount: Uint8Array; // 8 bytes
  commitment: Uint8Array; // 32 bytes
  leafIndex: number;
  tokenId?: Uint8Array; // 32 bytes (present for deposit/unshield, zero for private transfers)
}

/** Parsed sender memo event (Phase 2). */
export interface SenderMemoEvent {
  type: "sender_memo";
  /** 24-byte XChaCha20 nonce. */
  nonce: Uint8Array;
  /** 56-byte ChaCha20 ciphertext + Poly1305 tag. */
  ciphertextWithTag: Uint8Array;
  /** Commitment of the output this memo covers (also AAD). */
  commitment: Uint8Array;
  /** Leaf index of the covered output (also AAD). */
  leafIndex: number;
}

/**
 * BTC origin attestation, emitted alongside every SPV-verified deposit.
 * Lets third-party auditors anchor commitments to their on-chain BTC
 * origin without trusting our backend.
 *
 * Layout matches the Rust `emit_btc_origin_attestation` in
 * `contracts/programs/utxopia/src/utils/events.rs`.
 */
export interface BtcOriginAttestationEvent {
  type: "btc_origin_attestation";
  blockHeight: bigint;
  /** Bitcoin deposit txid in internal byte order (same as `complete_deposit` instruction data). */
  depositTxid: Uint8Array;
  /** Sweep transaction's output index that paid the pool. */
  sweepVout: number;
  /** Commitment inserted into the JoinSplit tree for this deposit. */
  commitment: Uint8Array;
  /** Pool-received amount in satoshis (after sweep fees). */
  amountSats: bigint;
}

export type ProgramEvent =
  | NullifierSpentEvent
  | StealthAnnouncementEvent
  | SenderMemoEvent
  | BtcOriginAttestationEvent;

/**
 * Parse a nullifier spent event from decoded sol_log_data segments.
 * Expected: disc(1) + nullifier_hash(32) + op_type(1)
 */
export function parseNullifierSpentEvent(segments: Uint8Array[]): NullifierSpentEvent | null {
  if (segments.length < 3) return null;
  if (segments[0].length !== 1 || segments[0][0] !== EVENT_NULLIFIER_SPENT) return null;

  const nullifierHash = segments[1];
  if (nullifierHash.length !== 32) return null;

  const opType = segments[2];
  if (opType.length !== 1) return null;

  return {
    type: "nullifier_spent",
    nullifierHash,
    operationType: opType[0],
  };
}

/**
 * Parse a stealth announcement event from decoded sol_log_data segments.
 * v1: disc(1) + type(1) + ephemeral_pub(32) + encrypted_amount(8) + commitment(32) + leaf_index(4) = 6 segments
 * v2: + token_id(32) = 7 segments
 */
export function parseStealthAnnouncementEvent(segments: Uint8Array[]): StealthAnnouncementEvent | null {
  if (segments.length < 6) return null;
  if (segments[0].length !== 1 || segments[0][0] !== EVENT_STEALTH_ANNOUNCEMENT) return null;

  const atype = segments[1];
  if (atype.length !== 1) return null;

  const ephemeralPub = segments[2];
  if (ephemeralPub.length !== 32) return null;

  const encryptedAmount = segments[3];
  if (encryptedAmount.length !== 8) return null;

  const commitment = segments[4];
  if (commitment.length !== 32) return null;

  const liBytes = segments[5];
  if (liBytes.length !== 4) return null;
  const view = new DataView(liBytes.buffer, liBytes.byteOffset, 4);
  const leafIndex = view.getUint32(0, true);

  // v2: token_id at segment 6
  let tokenId: Uint8Array | undefined;
  if (segments.length >= 7 && segments[6].length === 32) {
    tokenId = segments[6];
  }

  return {
    type: "stealth_announcement",
    announcementType: atype[0],
    ephemeralPub,
    encryptedAmount,
    commitment,
    leafIndex,
    tokenId,
  };
}

/**
 * Parse an association-set update event (Phase 3) from decoded sol_log_data segments.
 * Layout: disc(1) + new_root(32) + status(1) + version_le(8)
 */
/**
 * Parse a BTC origin attestation event from decoded sol_log_data segments.
 * Layout: disc(1) + block_height(8 LE) + deposit_txid(32) + sweep_vout(4 LE)
 *       + commitment(32) + amount_sats(8 LE)
 */
export function parseBtcOriginAttestationEvent(
  segments: Uint8Array[],
): BtcOriginAttestationEvent | null {
  if (segments.length < 6) return null;
  if (segments[0].length !== 1 || segments[0][0] !== EVENT_BTC_ORIGIN_ATTESTATION) return null;

  const bhBytes = segments[1];
  if (bhBytes.length !== 8) return null;
  let blockHeight = 0n;
  for (let i = 7; i >= 0; i--) blockHeight = (blockHeight << 8n) | BigInt(bhBytes[i]);

  const depositTxid = segments[2];
  if (depositTxid.length !== 32) return null;

  const voutBytes = segments[3];
  if (voutBytes.length !== 4) return null;
  const sweepVout = new DataView(
    voutBytes.buffer,
    voutBytes.byteOffset,
    4,
  ).getUint32(0, true);

  const commitment = segments[4];
  if (commitment.length !== 32) return null;

  const amtBytes = segments[5];
  if (amtBytes.length !== 8) return null;
  let amountSats = 0n;
  for (let i = 7; i >= 0; i--) amountSats = (amountSats << 8n) | BigInt(amtBytes[i]);

  return {
    type: "btc_origin_attestation",
    blockHeight,
    depositTxid,
    sweepVout,
    commitment,
    amountSats,
  };
}

/**
 * Parse a sender memo event (Phase 2) from decoded sol_log_data segments.
 * Layout: disc(1) + nonce(24) + ciphertext_and_tag(56) + commitment(32) + leaf_index(4)
 */
export function parseSenderMemoEvent(segments: Uint8Array[]): SenderMemoEvent | null {
  if (segments.length < 5) return null;
  if (segments[0].length !== 1 || segments[0][0] !== EVENT_SENDER_MEMO) return null;

  const nonce = segments[1];
  if (nonce.length !== 24) return null;

  const ciphertextWithTag = segments[2];
  if (ciphertextWithTag.length !== 56) return null;

  const commitment = segments[3];
  if (commitment.length !== 32) return null;

  const liBytes = segments[4];
  if (liBytes.length !== 4) return null;
  const leafIndex = new DataView(liBytes.buffer, liBytes.byteOffset, 4).getUint32(0, true);

  return { type: "sender_memo", nonce, ciphertextWithTag, commitment, leafIndex };
}

/**
 * Parse batched nullifiers from a single flat segment.
 * Layout: disc(1) + count(1) + op_type(1) + [hash(32)] x count
 */
function parseNullifiersBatch(data: Uint8Array): NullifierSpentEvent[] {
  if (data.length < 3) return [];
  const count = data[1];
  const opType = data[2];
  const expectedLen = 3 + count * 32;
  if (data.length < expectedLen) return [];

  const events: NullifierSpentEvent[] = [];
  for (let i = 0; i < count; i++) {
    const offset = 3 + i * 32;
    events.push({
      type: "nullifier_spent",
      nullifierHash: data.slice(offset, offset + 32),
      operationType: opType,
    });
  }
  return events;
}

/**
 * Parse batched announcements from a single flat segment.
 * v1: disc(1) + count(1) + [type(1) + ephemeral(32) + amount(8) + commitment(32) + leaf_index(4)] x count (77 per item)
 * v2: disc(1) + count(1) + [type(1) + ephemeral(32) + amount(8) + commitment(32) + leaf_index(4) + token_id(32)] x count (109 per item)
 */
function parseAnnouncementsBatch(data: Uint8Array): StealthAnnouncementEvent[] {
  if (data.length < 2) return [];
  const count = data[1];
  if (count === 0) return [];

  // Detect v1 vs v2 by checking total size
  const remainingBytes = data.length - 2;
  const v2ItemSize = 109;
  const v1ItemSize = 77;
  const isV2 = remainingBytes >= count * v2ItemSize;
  const itemSize = isV2 ? v2ItemSize : v1ItemSize;

  const expectedLen = 2 + count * itemSize;
  if (data.length < expectedLen) return [];

  const events: StealthAnnouncementEvent[] = [];
  for (let i = 0; i < count; i++) {
    const offset = 2 + i * itemSize;
    const liView = new DataView(data.buffer, data.byteOffset + offset + 73, 4);
    const event: StealthAnnouncementEvent = {
      type: "stealth_announcement",
      announcementType: data[offset],
      ephemeralPub: data.slice(offset + 1, offset + 33),
      encryptedAmount: data.slice(offset + 33, offset + 41),
      commitment: data.slice(offset + 41, offset + 73),
      leafIndex: liView.getUint32(0, true),
    };
    if (isV2) {
      event.tokenId = data.slice(offset + 77, offset + 109);
    }
    events.push(event);
  }
  return events;
}

function decodeBase64(b64: string): Uint8Array {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

/**
 * Parse program events from Solana transaction log messages.
 *
 * sol_log_data emits log lines in the format:
 *   "Program data: <base64_segment1> <base64_segment2> ..."
 *
 * @param logs - Array of log message strings from a transaction
 * @param programId - Optional program ID to filter events (matches "Program <id> invoke" blocks)
 */
export function parseProgramEvents(logs: string[], programId?: string): ProgramEvent[] {
  const events: ProgramEvent[] = [];
  const DATA_PREFIX = "Program data: ";

  for (const line of logs) {
    if (!line.startsWith(DATA_PREFIX)) continue;

    const b64Parts = line.slice(DATA_PREFIX.length).split(" ");
    const segments = b64Parts.map(decodeBase64);

    if (segments.length === 0) continue;

    // Handle batch events (single flat segment)
    if (segments.length === 1 && segments[0].length > 1) {
      const disc = segments[0][0];
      if (disc === EVENT_NULLIFIERS_BATCH) {
        events.push(...parseNullifiersBatch(segments[0]));
        continue;
      }
      if (disc === EVENT_ANNOUNCEMENTS_BATCH) {
        events.push(...parseAnnouncementsBatch(segments[0]));
        continue;
      }
    }

    if (segments[0].length !== 1) continue;

    const disc = segments[0][0];

    if (disc === EVENT_NULLIFIER_SPENT) {
      const event = parseNullifierSpentEvent(segments);
      if (event) events.push(event);
    } else if (disc === EVENT_STEALTH_ANNOUNCEMENT) {
      const event = parseStealthAnnouncementEvent(segments);
      if (event) events.push(event);
    } else if (disc === EVENT_SENDER_MEMO) {
      const event = parseSenderMemoEvent(segments);
      if (event) events.push(event);
    } else if (disc === EVENT_BTC_ORIGIN_ATTESTATION) {
      const event = parseBtcOriginAttestationEvent(segments);
      if (event) events.push(event);
    }
  }

  return events;
}

File: contracts/programs/utxopia/src/utils/events.rs
//! Event emission utilities using sol_log_data
//!
//! Events are emitted as base64-encoded log lines ("Program data: <base64>")
//! and can be parsed by the backend indexer.
//!
//! ## Events
//!
//! - 0x02 NullifierSpent: disc(1) + nullifier_hash(32) + op_type(1) + ix_disc(1) = 35 bytes
//! - 0x03 StealthAnnouncement: disc(1) + type(1) + ephemeral(32) + amount(8) + commitment(32) + leaf_index(4) + token_id(32) = 110 bytes
//! - 0x07 RedemptionCompleted: variable
//! - 0x08 RedemptionRequested: variable
//! - 0x0A RedemptionProcessing: disc(1) + requester(32) + amount(8) + request_id(8) + slot(4) = 53 bytes
//! - 0x0B NullifiersBatch: disc(1) + count(1) + op_type(1) + ix_disc(1) + [hash(32)] x N
//! - 0x0C AnnouncementsBatch: disc(1) + count(1) + [type(1) + ephemeral(32) + amount(8) + commitment(32) + leaf_index(4)] x N
//! - 0x0D DepositVerified: disc(1) + sweep_txid(32) + deposit_txid(32) + amount_sats(8) + leaf_index(4) + original_amount(8) = 85 bytes
//! - 0x0E UnshieldMeta: disc(1) + gross_amount(8) + fee(8) + payout(8) + recipient(32) + token_id(32) = 89 bytes
//! - 0x0F UtxoCreated: disc(1) + txid(32) + vout(4) + amount_sats(8) = 45 bytes
//! - 0x10 UtxoConsumed: disc(1) + txid(32) + vout(4) + amount_sats(8) = 45 bytes
//! - 0x11 ShieldMeta: disc(1) + gross_amount(8) + fee(8) + token_id(32) = 49 bytes

use pinocchio::log::sol_log_data;

/// Event discriminator: nullifier spent
const EVENT_NULLIFIER_SPENT: u8 = 0x02;

/// Event discriminator: stealth announcement (replaces on-chain PDA)
const EVENT_STEALTH_ANNOUNCEMENT: u8 = 0x03;

/// Announcement type: deposit (plaintext amount from BTC deposit verification)
pub const ANNOUNCEMENT_TYPE_DEPOSIT: u8 = 0;

/// Announcement type: transfer (XOR-encrypted amount from JoinSplit transact)
pub const ANNOUNCEMENT_TYPE_TRANSFER: u8 = 1;

/// Event discriminator: redemption completed (PDA about to close)
const EVENT_REDEMPTION_COMPLETED: u8 = 0x07;

/// Event discriminator: redemption requested (PDA created)
const EVENT_REDEMPTION_REQUESTED: u8 = 0x08;

/// Event discriminator: redemption marked as processing
const EVENT_REDEMPTION_PROCESSING: u8 = 0x0A;

/// Event discriminator: batched nullifiers spent
const EVENT_NULLIFIERS_BATCH: u8 = 0x0B;

/// Event discriminator: batched stealth announcements
const EVENT_ANNOUNCEMENTS_BATCH: u8 = 0x0C;

/// Event discriminator: deposit verified via SPV (carries BTC txids + amount)
const EVENT_DEPOSIT_VERIFIED: u8 = 0x0D;

/// Event discriminator: unshield/redeem metadata (amount + recipient)
const EVENT_UNSHIELD_META: u8 = 0x0E;

/// Event discriminator: UTXO created (deposit or change)
const EVENT_UTXO_CREATED: u8 = 0x0F;

/// Event discriminator: UTXO consumed (spent in withdrawal)
const EVENT_UTXO_CONSUMED: u8 = 0x10;

/// Event discriminator: shield metadata (gross amount + fee for SPL/BTC deposits)
const EVENT_SHIELD_META: u8 = 0x11;

/// Event discriminator: sender memo (Phase 2 — outgoing visibility for the
/// sender's own viewing key). Encrypted with XChaCha20-Poly1305 AEAD using
/// an `ovk = SHA-256(viewingPrivKey || "utxopia.ovk.v1")` outgoing viewing
/// key. AAD = commitment || leaf_index_le, so the Poly1305 tag binds the
/// memo to its tree leaf (regulator-grade tamper detection).
///
/// Layout:
///   disc(1) + nonce(24) + ciphertext_and_tag(56) + commitment(32) + leaf_index(4)
///   = 117 bytes
pub const EVENT_SENDER_MEMO: u8 = 0x12;

// Event discriminators 0x13 (ASSOCIATION_ROOT_UPDATED) and 0x14 (POI_ATTESTED)
// were previously PoI machinery. Removed alongside the on-chain PoI instructions.

/// Event discriminator: BTC deposit origin attestation.
///
/// Emitted alongside `EVENT_DEPOSIT_VERIFIED` for every successful
/// `complete_deposit`. Complements PoI by giving third-party auditors
/// the raw BTC origin data they need to build their own association sets
/// without trusting our backend. Anyone running an indexer can subscribe
/// to disc 0x15 events, validate them against the chain (the program
/// already enforced SPV before emitting), and aggregate the commitments.
///
/// Layout: disc(1) + block_height(8 LE) + deposit_txid(32, internal order)
///       + sweep_vout(4 LE) + commitment(32) + amount_sats(8 LE) = 85 bytes
pub const EVENT_BTC_ORIGIN_ATTESTATION: u8 = 0x15;

// Event discriminators 0x16 (POI_HIDDEN_ATTESTED) and 0x17 (TRANSACT_WITH_POI)
// were PoI machinery, now removed. The compliance event stream is built on
// disc 0x15 (EVENT_BTC_ORIGIN_ATTESTATION) plus a planned 0x18
// (EVENT_ORIGIN_SCREENED) for passive attestation — see docs/COMPLIANCE.md.

/// Max batch items for stack-allocated buffer (MAX_SAFE_JOINSPLIT_SIZE = 14)
const MAX_BATCH: usize = 14;

/// Emit when a nullifier is spent.
///
/// Layout: disc(1) + nullifier_hash(32) + op_type(1) + ix_disc(1) = 35 bytes
pub fn emit_nullifier_spent(
    nullifier_hash: &[u8; 32],
    operation_type: u8,
    instruction_disc: u8,
) {
    let disc = [EVENT_NULLIFIER_SPENT];
    let op = [operation_type];
    let ix = [instruction_disc];
    sol_log_data(&[&disc, nullifier_hash.as_ref(), &op, &ix]);
}

/// Emit a stealth announcement with token_id.
///
/// Layout: disc(1) + type(1) + ephemeral_pub(32) + encrypted_amount(8)
///         + commitment(32) + leaf_index(4) + token_id(32) = 110 bytes
pub fn emit_stealth_announcement(
    announcement_type: u8,
    ephemeral_pub: &[u8; 32],
    encrypted_amount: &[u8; 8],
    commitment: &[u8; 32],
    leaf_index: u32,
    token_id: &[u8; 32],
) {
    let disc = [EVENT_STEALTH_ANNOUNCEMENT];
    let atype = [announcement_type];
    let li = leaf_index.to_le_bytes();
    sol_log_data(&[&disc, &atype, ephemeral_pub, encrypted_amount, commitment, &li, token_id]);
}

/// Emit when a redemption is completed (before PDA is closed).
///
/// Layout: disc(1) + requester(32) + amount_sats(8) + actual_received(8) + service_fee(8)
///         + request_id(8) + btc_txid(32) + btc_script_len(1) + btc_script(var)
/// = 114 + btc_script_len bytes
///
/// Trustless accounting:
///   miner_fee        = total_input_sats - sum(tx_outputs)  [computed on-chain from raw BTC tx]
///   burn_amount      = actual_received + miner_fee          [BTC that left the pool]
///   protocol_revenue = service_fee - miner_fee              [net profit kept in vault]
pub fn emit_redemption_completed(
    requester: &[u8; 32],
    amount_sats: u64,
    actual_received: u64,
    service_fee: u64,
    request_id: u64,
    btc_txid: &[u8; 32],
    burn_amount: u64,
    protocol_revenue: u64,
    btc_script: &[u8],
) {
    let disc = [EVENT_REDEMPTION_COMPLETED];
    let amt = amount_sats.to_le_bytes();
    let recv = actual_received.to_le_bytes();
    let sfee = service_fee.to_le_bytes();
    let rid = request_id.to_le_bytes();
    let burn_bytes = burn_amount.to_le_bytes();
    let proto_bytes = protocol_revenue.to_le_bytes();
    let script_len = [btc_script.len() as u8];
    sol_log_data(&[&disc, requester, &amt, &recv, &sfee, &rid, btc_txid, &burn_bytes, &proto_bytes, &script_len, btc_script]);
}

/// Emit when a redemption request is created (PDA initialized).
///
/// Layout: disc(1) + requester(32) + amount_sats(8) + request_id(8)
///         + service_fee_base(8) + service_fee_bps(2)
///         + btc_script_len(1) + btc_script(var)
pub fn emit_redemption_requested(
    requester: &[u8; 32],
    amount_sats: u64,
    request_id: u64,
    service_fee_base: u64,
    service_fee_bps: u16,
    btc_script: &[u8],
) {
    let disc = [EVENT_REDEMPTION_REQUESTED];
    let amt = amount_sats.to_le_bytes();
    let rid = request_id.to_le_bytes();
    let sfb = service_fee_base.to_le_bytes();
    let sbps = service_fee_bps.to_le_bytes();
    let script_len = [btc_script.len() as u8];
    sol_log_data(&[&disc, requester, &amt, &rid, &sfb, &sbps, &script_len, btc_script]);
}

/// Emit when a redemption transitions to Processing state.
///
/// Layout: disc(1) + requester(32) + amount_sats(8) + request_id(8) + processing_slot(4)
pub fn emit_redemption_processing(
    requester: &[u8; 32],
    amount_sats: u64,
    request_id: u64,
    processing_slot: u32,
) {
    let disc = [EVENT_REDEMPTION_PROCESSING];
    let amt = amount_sats.to_le_bytes();
    let rid = request_id.to_le_bytes();
    let slot = processing_slot.to_le_bytes();
    sol_log_data(&[&disc, requester, &amt, &rid, &slot]);
}

/// Emit a batch of nullifier spent events in a single sol_log_data call.
///
/// Layout: disc(1) + count(1) + op_type(1) + ix_disc(1) + [nullifier_hash(32)] x count
pub fn emit_nullifiers_batch(
    nullifiers: &[&[u8; 32]],
    operation_type: u8,
    instruction_disc: u8,
) {
    // For single nullifier, use the non-batch version (simpler parsing)
    if nullifiers.len() == 1 {
        emit_nullifier_spent(nullifiers[0], operation_type, instruction_disc);
        return;
    }

    let disc = [EVENT_NULLIFIERS_BATCH];
    let count = [nullifiers.len() as u8];
    let op = [operation_type];
    let ix = [instruction_disc];

    // Build slice array: disc + count + op_type + ix_disc + N hashes
    // Max slices = 4 + MAX_BATCH = 18
    let mut slices: [&[u8]; 4 + MAX_BATCH] = [&[0u8; 0]; 4 + MAX_BATCH];
    slices[0] = &disc;
    slices[1] = &count;
    slices[2] = &op;
    slices[3] = &ix;
    let n = nullifiers.len().min(MAX_BATCH);
    for i in 0..n {
        slices[4 + i] = nullifiers[i].as_ref();
    }
    sol_log_data(&slices[..4 + n]);
}

/// Emit when a BTC deposit is SPV-verified on-chain.
///
/// Layout: disc(1) + sweep_txid(32) + deposit_txid(32) + amount_sats(8) + leaf_index(4) + original_amount(8) = 85 bytes
pub fn emit_deposit_verified(
    sweep_txid: &[u8; 32],
    deposit_txid: &[u8; 32],
    amount_sats: u64,
    leaf_index: u32,
    original_amount: u64,
) {
    let disc = [EVENT_DEPOSIT_VERIFIED];
    let amt = amount_sats.to_le_bytes();
    let li = leaf_index.to_le_bytes();
    let orig = original_amount.to_le_bytes();
    sol_log_data(&[&disc, sweep_txid, deposit_txid, &amt, &li, &orig]);
}

/// Emit a BTC origin attestation for every SPV-verified deposit.
///
/// Layout: disc(1) + block_height(8 LE) + deposit_txid(32) + sweep_vout(4 LE)
///       + commitment(32) + amount_sats(8 LE) = 85 bytes.
///
/// This complements `EVENT_DEPOSIT_VERIFIED` by including the on-chain
/// commitment + sweep output index, so a third-party auditor can build an
/// association set keyed on commitment hashes without re-deriving them.
pub fn emit_btc_origin_attestation(
    block_height: u64,
    deposit_txid: &[u8; 32],
    sweep_vout: u32,
    commitment: &[u8; 32],
    amount_sats: u64,
) {
    let disc = [EVENT_BTC_ORIGIN_ATTESTATION];
    let bh = block_height.to_le_bytes();
    let vout = sweep_vout.to_le_bytes();
    let amt = amount_sats.to_le_bytes();
    sol_log_data(&[&disc, &bh, deposit_txid, &vout, commitment, &amt]);
}

/// Emit unshield/redeem metadata so indexer doesn't need to parse instruction data.
///
/// Layout: disc(1) + gross_amount(8) + fee(8) + payout(8) + recipient(32) + token_id(32) = 89 bytes
pub fn emit_unshield_meta(
    gross_amount: u64,
    fee: u64,
    payout: u64,
    recipient: &[u8; 32],
    token_id: &[u8; 32],
) {
    let disc = [EVENT_UNSHIELD_META];
    let gross = gross_amount.to_le_bytes();
    let f = fee.to_le_bytes();
    let p = payout.to_le_bytes();
    sol_log_data(&[&disc, &gross, &f, &p, recipient, token_id]);
}

/// Emit when a UTXO is created (deposit or change output).
///
/// Layout: disc(1) + txid(32) + vout(4) + amount_sats(8) = 45 bytes
pub fn emit_utxo_created(
    txid: &[u8; 32],
    vout: u32,
    amount_sats: u64,
) {
    let disc = [EVENT_UTXO_CREATED];
    let v = vout.to_le_bytes();
    let amt = amount_sats.to_le_bytes();
    sol_log_data(&[&disc, txid, &v, &amt]);
}

/// Emit when a UTXO is consumed (spent in a withdrawal tx).
///
/// Layout: disc(1) + txid(32) + vout(4) + amount_sats(8) = 45 bytes
pub fn emit_utxo_consumed(
    txid: &[u8; 32],
    vout: u32,
    amount_sats: u64,
) {
    let disc = [EVENT_UTXO_CONSUMED];
    let v = vout.to_le_bytes();
    let amt = amount_sats.to_le_bytes();
    sol_log_data(&[&disc, txid, &v, &amt]);
}

/// Emit shield metadata so indexer can record gross amount and fee.
///
/// Layout: disc(1) + gross_amount(8) + fee(8) + token_id(32) = 49 bytes
pub fn emit_shield_meta(
    gross_amount: u64,
    fee: u64,
    token_id: &[u8; 32],
) {
    let disc = [EVENT_SHIELD_META];
    let gross = gross_amount.to_le_bytes();
    let f = fee.to_le_bytes();
    sol_log_data(&[&disc, &gross, &f, token_id]);
}

/// Emit a sender memo (Phase 2 — outgoing visibility for the sender's own viewing key).
///
/// Encrypted with XChaCha20-Poly1305 AEAD using `ovk` derived from the sender's
/// viewing private key. The Poly1305 tag is included in `ciphertext_and_tag`
/// (last 16 bytes); AAD = `commitment || leaf_index_le` so the tag binds the
/// memo to its tree leaf.
///
/// Layout: disc(1) + nonce(24) + ciphertext_and_tag(56) + commitment(32) + leaf_index(4)
///         = 117 bytes
pub fn emit_sender_memo(
    nonce: &[u8; 24],
    ciphertext_and_tag: &[u8; 56],
    commitment: &[u8; 32],
    leaf_index: u32,
) {
    let disc = [EVENT_SENDER_MEMO];
    let li = leaf_index.to_le_bytes();
    sol_log_data(&[&disc, nonce, ciphertext_and_tag, commitment, &li]);
}

/// Data for a single announcement in a batch (with token_id)
pub struct AnnouncementItem<'a> {
    pub announcement_type: u8,
    pub ephemeral_pub: &'a [u8; 32],
    pub encrypted_amount: &'a [u8; 8],
    pub commitment: &'a [u8; 32],
    pub leaf_index: u32,
    pub token_id: &'a [u8; 32],
}

/// Emit a batch of stealth announcements in a single sol_log_data call.
///
/// Layout: disc(1) + count(1) + [type(1) + ephemeral(32) + amount(8) + commitment(32) + leaf_index(4) + token_id(32)] x count
/// Per-item: 109 bytes. Max payload: 2 + 14 * 109 = 1528 bytes.
pub fn emit_announcements_batch(items: &[AnnouncementItem]) {
    if items.len() == 1 {
        emit_stealth_announcement(
            items[0].announcement_type,
            items[0].ephemeral_pub,
            items[0].encrypted_amount,
            items[0].commitment,
            items[0].leaf_index,
            items[0].token_id,
        );
        return;
    }

    let n = items.len().min(MAX_BATCH);

    // Max payload: 2 + 14 * 109 = 1528 bytes — fits on stack
    let mut buf = [0u8; 2 + MAX_BATCH * 109];
    buf[0] = EVENT_ANNOUNCEMENTS_BATCH;
    buf[1] = n as u8;
    let mut offset = 2;
    for i in 0..n {
        buf[offset] = items[i].announcement_type;
        offset += 1;
        buf[offset..offset + 32].copy_from_slice(items[i].ephemeral_pub);
        offset += 32;
        buf[offset..offset + 8].copy_from_slice(items[i].encrypted_amount);
        offset += 8;
        buf[offset..offset + 32].copy_from_slice(items[i].commitment);
        offset += 32;
        let li = items[i].leaf_index.to_le_bytes();
        buf[offset..offset + 4].copy_from_slice(&li);
        offset += 4;
        buf[offset..offset + 32].copy_from_slice(items[i].token_id);
        offset += 32;
    }

    sol_log_data(&[&buf[..offset]]);
}
